Log to syslog whether the DP is online or offline

fidencio opened the pull request:
This PR is supposed to cover a small bit of https://pagure.io/SSSD/sssd/issue/3155 and log to syslog whether the DP is online or offline.

Hopefully by logging this to syslog it'll help admins to find out those issues in a simpler way as nowadays they'd have to enable the logs, search for this info there ... which is not exactly handy.

fidencio commented:
I've updated the patches with the link to the right pagure issue.

fidencio commented:
Any taker?

jhrozek commented on src/providers/data_provider_be.c:
```diff
@@ -249,7 +249,7 @@ static errno_t be_check_online_request(struct be_ctx *be_ctx)
 static void be_check_online_done(struct tevent_req *req)
 {
     struct be_ctx *be_ctx;
-    struct dp_reply_std reply;
```
I guess we never caught this bug because of the (void**) cast, but I wonder how did you find this bug? Did it cause any issues in the code? If yes,it might be a good candidate for a backport.

fidencio replied:
I found this bug while debugging why the "Backend is Online" and "Backend is Offline" messages were never triggered. Apart from this small issue, I'd say reset_fo() was being called way more times than actually needed.

So, yes, probably this patch could be backported.

pbrezina commented:
It would be nice also to include domain name in the syslog message.

fidencio commented:
@pbrezina, you mean for covering cases like subdomains?

Because it's already printed with the domain name (as it comes from the domain process).

jhrozek commented on src/providers/data_provider_be.c:
```diff
@@ -249,8 +249,9 @@ static errno_t be_check_online_request(struct be_ctx *be_ctx)
-    struct dp_reply_std reply;
+    struct dp_reply_std *reply;
     errno_t ret;
+    static int last_dp_state = -1;
```
I'm abvivalent about a static variable, because I don't think functions should have state, but I guess in the end it is a bit cleaner than putting another field in the be context as long as this variable is only used for logging, but could you please add a comment that this is only about logging so that nobody tries to optimize the code in the future?

fidencio replied:
Well, I can either add this state to the be_context or a comment to the static var.
Just let me know what's your preference.

jhrozek commented on src/providers/data_provider_be.c:
```diff
@@ -260,11 +261,19 @@ static void be_check_online_done(struct tevent_req *req)
     case DP_ERR_OK:
+        if (last_dp_state != DP_ERR_OK) {
+            last_dp_state = DP_ERR_OK;
+            sss_log(SSS_LOG_INFO, "Backend is online\n");
```
Does SSS_LOG_INFO turn up in the syslog by default on RHEL-7?

fidencio replied:
Hmm. That's a good question, I'll check it soon and change it in case it's needed.

lslebodn replied:
What is a answer to Jakub's question?

fidencio replied:
@lslebodn, you have missed https://github.com/SSSD/sssd/pull/404#issuecomment-336918930

There I've said (and I'm quoting here): "tested with RHEL-7 and confirmed that SSS_LOG_INFO is enough"

fidencio commented:
Adding "Changes Requested" label per @pbrezina and @jhrozek reviews.

fidencio commented:
@pbrezina, @jhrozek: Another question is ... is still worh to keep the debug message at SSSDBG_TRACE_FUNC level that is exactly like the one printed in the syslog?

fidencio commented:
Answering my own question, IMO doesn't make sense to remove the message as it would not be handy to figure out what's happening just by the logs without those. OTOH, I'll print those only when the state changes.

jhrozek commented:
Yes, I also prefer to have the message in the debug logs as well.

fidencio commented:
Okay, I've updated the PR:
- rebased atop of master;
- changed the static variable to another field in the be_ctx struct;
- tested with RHEL-7 and confirmed that SSS_LOG_INFO is enough

@pbrezina, I haven't added the domain in the syslog message as it seems to be redundant (as printing be_ctx->domain->name wouldn't add any new info in the following printed message: `Oct 16 16:58:01 client.debug.fidencio sssd[be[debug.fidencio]][6330]: Backend is offline`)

I've added a comment in the code that has the purpose to remind us to add the domain to those logs whenever we have the backend status per domain.

pbrezina commented:
Ack to those patches.

BTW I think we can drop the last patch adding the comment about including domain name in the future.

fidencio commented:
I've pushed a new version of the patchset, just dropping the last patch which added the comment about including domai name in the future, as per @pbrezina's request.

As any other change has been done, I'm keeping the "Accepted" label added by @pbrezina.

lslebodn commented:
Please explain why do you think that last patch is a good idea "DP: Only print its status in case of change"

fidencio commented:
@lslebodn: Because I don't see the point of having *a* *lot* of consecutive patches saying "Backend is offline" for instance.

But if you disagree, I can just drop the patch.

Please, as this is something desired for the release on Friday I'd like to ask for a quick answer.

lslebodn commented:
On (18/10/17 04:24), fidencio wrote:
>fidencio commented on this pull request.
>
>
>
>>      case DP_ERR_OK:
>+        if (last_dp_state != DP_ERR_OK) {
>+            last_dp_state = DP_ERR_OK;
>+            sss_log(SSS_LOG_INFO, "Backend is online\n");
>
>@lslebodn, you have missed https://github.com/SSSD/sssd/pull/404#issuecomment-336918930
>
>There I've said (and I'm quoting here): "tested with RHEL-7 and confirmed that SSS_LOG_INFO is enough"
>

This an issue with mixing inline comments with ordinary one.

LS

lslebodn commented:
On (18/10/17 04:22), fidencio wrote:
>@lslebodn: Because I don't see the point of having *a* *lot* of consecutive patches saying "Backend is offline" for instance.
>
>But if you disagree, I can just drop the patch.
>

Let's imagine that you will analyze 100MiB file.
And you are not sure about state of backend because last change was
17469 lines before.

Tracing debug messages have purpose.

>Please, as this is something desired for the release on Friday I'd like to ask for a quick answer.
>

And I prefer quality instead of half baked features before deadliens
(but it is not this case :-)

LS

fidencio commented:
On Wed, Oct 18, 2017 at 2:34 PM, lslebodn <notifications@github.com> wrote:

> On (18/10/17 04:22), fidencio wrote:
> >@lslebodn: Because I don't see the point of having *a* *lot* of
> consecutive patches saying "Backend is offline" for instance.
> >
> >But if you disagree, I can just drop the patch.
> >
>
> Let's imagine that you will analyze 100MiB file.
> And you are not sure about state of backend because last change was
> 17469 lines before.
>
> Tracing debug messages have purpose.
>

Okay, makes sense. Patch has been dropped from the PR.


>
> >Please, as this is something desired for the release on Friday I'd like
> to ask for a quick answer.
> >
>
> And I prefer quality instead of half baked features before deadliens
> (but it is not this case :-)


Can you elaborate a bit more in the "but it is not this case" part of
sentence?

I'm basically interested to know whether you meant that this PR is an
example of a non-quality half-baked feature. So, can you tell me whether
that's what you meant?


>
>
> LS
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SSSD/sssd/pull/404#issuecomment-337575604>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAG4emLa8YT72Tpt1m_GRI6WazSwVzqtks5stfBFgaJpZM4P2q30>
> .
>

lslebodn commented:
On (18/10/17 12:41), fidencio wrote:
>> And I prefer quality instead of half baked features before deadliens
>> (but it is not this case :-)
>
>
>Can you elaborate a bit more in the "but it is not this case" part of
>sentence?
>
>I'm basically interested to know whether you meant that this PR is an
>example of a non-quality half-baked feature. So, can you tell me whether
>that's what you meant?
>

I meant that this PR is not half baked feature. Sorry for confusion.

LS

lslebodn commented:
master:
* 3542fe821765cad1f25f6c2a077b55fc1d7d0553
* 7f95edc43d9fc410aab5712552e17f28932ba344
